### ISSUE TITLE
Corruption of SRTP sequence numbers leads to "Discarded invalid SRTP packet: authentication failed" and no further audio

### ISSUE THREAD

**Issue by elbow** [sipwise/rtpengine#955]
Hi,

I know that this symptom has come up a number of times over the years, but I have a specific example which I've investigated thoroughly.

We have many customers who connect into our core network using a proprietary "voice optimising" VPN product called Vibe.  This product trunks packets giving rtp absolute first preference.  At the same time it compresses RTP headers which obviously requires some heuristics/guesswork to reinstate the header on the receiving side.

Occasionally when the en-vibing router drops some packets on the way into the vibe process, we see that on the receiving end the sequence number does not skip but instead continues as if there were no missed packets.  The timestamp, however, jumps ahead.

This is obviously wrong but the impact seems to be minor with ordinary RTP and a tolerant receiver (its Asterisk 13 for us).

But we are using SRTP and relying on rtpengine to decrypt and (often) transcode.

In the example that I captured sequence numbers 9126 and 9127 were lost.  Vibe successfully did convey source packet 9128, but when it delivered it out at the other end it used sequence number 9126.  (and so on for 9129/9130 etc),   So the reconstructed sequence number "slips" by 2. The payload and the timestamp were delivered intact.

Here's a diff of the sent 9128 and the way it was delivered:

```
--- client-end-seq-9128.txt	2020-03-22 10:24:52.000000000 +0200
+++ host-end-seq-9126.txt	2020-03-22 10:11:42.000000000 +0200
@@ -1,5 +1,5 @@
 No.     Time                          Source                Destination           Protocol Length Info
-  13396 2018-04-27 05:13:30.067894    10.64.16.104          197.155.250.157       RTP      226    PT=ITU-T G.722, SSRC=0xD4B6FB0D, Seq=9128, Time=254363141
+  22981 2020-03-22 07:26:19.337289    10.64.16.104          197.155.250.157       RTP      226    PT=ITU-T G.722, SSRC=0xD4B6FB0D, Seq=9126, Time=254363141

-Frame 13396: 226 bytes on wire (1808 bits), 226 bytes captured (1808 bits)
+Frame 22981: 226 bytes on wire (1808 bits), 226 bytes captured (1808 bits)
 Linux cooked capture
@@ -8,4 +8,4 @@
     Link-layer address length: 6
-    Source: Routerbo_b4:70:bc (cc:2d:e0:b4:70:bc)
-    Unused: 4000
+    Source: fe:fd:00:00:00:00 (fe:fd:00:00:00:00)
+    Unused: 0000
     Protocol: IPv4 (0x0800)
@@ -16,8 +16,8 @@
     Total Length: 210
-    Identification: 0xf62d (63021)
+    Identification: 0xe961 (59745)
     Flags: 0x0000
     Fragment offset: 0
-    Time to live: 62
+    Time to live: 61
     Protocol: UDP (17)
-    Header checksum: 0xab0c [validation disabled]
+    Header checksum: 0xb8d8 [validation disabled]
     [Header checksum status: Unverified]
@@ -29,8 +29,8 @@
     Length: 190
-    Checksum: 0x0357 [unverified]
-    [Checksum Status: Unverified]
-    [Stream index: 3]
+    [Checksum: [missing]]
+    [Checksum Status: Not present]
+    [Stream index: 2]
     [Timestamps]
-        [Time since first frame: 45.713609000 seconds]
-        [Time since previous frame: 0.009603000 seconds]
+        [Time since first frame: 45.717321000 seconds]
+        [Time since previous frame: 0.011878000 seconds]
 Real-Time Transport Protocol
@@ -42,3 +42,3 @@
     Payload type: ITU-T G.722 (9)
-    Sequence number: 9128
+    Sequence number: 9126
     Timestamp: 254363141
```

Everything else of course was the same.

This is the point at which rtpengine freaked out.  Which is quite right, of course, since, from its point of view, it received and processed 9125, then the alleged 9126 arrived but the payload was not consistent.  

We did obviously report this problem to the Vibe developers and we await a response.

But my question is - is there no way that rtpengine can recover from this and continue?  Is there a flag somewhere that I missed that gives me some control over this?  (Obviously I don't want an ear-splitting squawk in the audio either...)

I have full traces and etc if they are required - rather not post them publicly though.

**Comment by rfuchs** [comment]
No, there is no such feature. Since the sequence number comes straight from the packet itself, it's expected to be correct. There is some heuristics in place for the ROC (the roll-over counter that extends the 16-bit sequence number to 32 bits) as the ROC doesn't appear in the packet itself but is being tracked implicitly and so it can happen that it goes out of sync (off by one, or perhaps reset to 0).

You can see the code for the ROC heuristics starting at https://github.com/sipwise/rtpengine/blob/master/daemon/rtp.c#L144 down to line 167. You can try to extend this code to not only try ROC mismatches (offsets of 0x10000) but also single sequence number offsets.

**Comment by elbow** [comment]
Hi Richard,

Thanks for the reply - I hope it was clear from my issue that its clear to me that this is the vibe team's issue and they did send some suggestions on their side.

I will look at the code but obviously I don't want to be responsible for adding fragility into rtpengine.

Steve

**Comment by rfuchs** [comment]
I would be ok with such a feature if it was made optional and disabled by default.

**Comment by systemcrash** [comment]
Isn't this a replay protection mechanism of SRTP?

SRTP reacts by dropping media which tries to re-use previously encountered sequence numbers. This vibe thing 'must' continue the conveyed sequence numbers for SRTP not to discard it(?).
